Add postgres source connector docs. (#323)
streamnative/pulsar-hub: connectors/debezium-postgres-source/3.0.1.6/debezium-postgres-source.md

---
description: The Debezium source connector pulls messages from PostgreSQL and persists the messages to Pulsar topics.
author: ["ASF"]
contributors: ["ASF"]
language: Java
document: 
source: "https://github.com/streamnative/pulsar/tree/v3.0.1.6/pulsar-io/debezium/postgres"
license: Apache License 2.0
tags: ["Pulsar IO", "Debezium", "PostgreSQL", "Postgres", "Source"]
alias: Debezium PostgreSQL Source
features: ["Use Debezium PostgreSQL source connector to sync data to Pulsar"]
license_link: "https://www.apache.org/licenses/LICENSE-2.0"
icon: "/images/connectors/debezium.jpg"
download: "https://github.com/streamnative/pulsar/releases/download/v3.0.1.6/pulsar-io-debezium-postgres-3.0.1.6.nar"
support: StreamNative
support_link: https://streamnative.io
support_img: "/images/streamnative.png"
owner_name: ""
owner_img: ""
dockerfile:
sn_available: true
id: "debezium-postgres-source"
---

The Postgres source connector pulls messages from PostgreSQL and persists the messages to Pulsar topics by using debezium.

![](/images/connectors/postgres-source.png)

## Quick start

### Prerequisites

> The prerequisites for connecting an Debezium Postgres source connector to external systems include:

Reviewer: ```suggestion
The prerequisites for connecting a Debezium Postgres source connector to external systems include:
```

- [Create a Connector with Function mesh](https://functionmesh.io/docs/connectors/run-connector).
  Using this way requires you to set the docker image. You can choose the version you want to launch from [here](https://hub.docker.com/r/streamnative/pulsar-io-elastic-search).

No matter how you create a Debezium Postgres source connector, the minimum connector configuration contains the following parameters:
```yaml
configs:
    database.hostname: {{Your hostname of Postgres}}
    database.port: {{Your port of Postgres}}
    database.user: {{Your user of Postgres}}
    database.password: {{Your password of Postgres}}
    database.dbname: {{Your dbname of Postgres}}
    # You can set multiple tables, and the connector will send data from each table to a different topic of pulsar, 
    # and the topic naming role is: {{database.server.name}}.{{table.name}}. For examples: "public/default/mydbserver.public.io-test"
    table.whitelist: "public.io-test" 
    database.server.name: "mydbserver"
    plugin.name: "pgoutput" 
```

> > * The configuration structure varies depending on how you create the AWS Kinesis sink connector.

Reviewer: ```suggestion
> * The configuration structure varies depending on how you create the Postgres source connector.
```

| `pulsar.service.url`                  | false    | null    | Pulsar cluster service URL.                                                                                                                                                                                                                                                |
| `offset.storage.topic`                | false    | null    | Record the last committed offsets that the connector successfully completes.                                                                                                                                                                                               |

## Advanced features

### Converter options

- org.apache.kafka.connect.json.JsonConverter

  The`json-with-envelope` config is valid only for the JsonConverter. By default, the value is set to false. When the `json-with-envelope` value is set to false, the consumer uses the schema `Schema.KeyValue(Schema.AUTO_CONSUME(), Schema.AUTO_CONSUME(), KeyValueEncodingType.SEPARATED)`, and the message only consists of the payload.

>   When the `json-with-envelope` value is set to true, the consumer uses the schema `Schema.KeyValue(Schema.BYTES, Schema.BYTES`, and the message consists of the schema and the payload.

Reviewer: ```suggestion
  When the `json-with-envelope` value is set to true, the consumer uses the schema `Schema.KeyValue(Schema.BYTES, Schema.BYTES)`, and the message consists of the schema and the payload.
```

- org.apache.pulsar.kafka.shade.io.confluent.connect.avro.AvroConverter

  If you select the AvroConverter, the consumer uses the schema `Schema.KeyValue(Schema.AUTO_CONSUME(), Schema.AUTO_CONSUME(), KeyValueEncodingType.SEPARATED)`, and the message consists of the payload.

### Used topic on Pulsar
Currently, the destination topic (specified by the `destination-topic-name` option ) is a required configuration but it is not used for the Debezium connector to save data. The Debezium connector saves data on the following 4 types of topics:

- One topic for storing the database metadata messages. It is named with the database server name ( `database.server.name`), like `public/default/database.server.name`.
- One topic (`offset.storage.topic`) for storing the offset metadata messages. The connector saves the last successfully-committed offsets on this topic.
- (Option) One topic (`database.history.pulsar.topic`) for storing the database history information. The connector writes and recovers DDL statements on this topic.

> - One per-table topic. The connector writes change events for all operations that occur in a table to a single Pulsar topic that is specific to that table.

Reviewer: The format for this topic is for example like: `mydbserver.public.io-test` right? Maybe it's better to add the topic name format here.
